File: frontend/src/utils/gpaPolicy.js
export const LEGACY_GPA_MODE = 'through_2024';
export const MODERN_GPA_MODE = 'from_2025';
export const GPA_MODE_OPTIONS = [
  { label: '2024级及以前', value: LEGACY_GPA_MODE },
  { label: '2025级及以后', value: MODERN_GPA_MODE },
];

export const gpaExclusionReason = (course, policy = {}) => {
  if (policy?.mode !== MODERN_GPA_MODE) return '';
  const code = course.code || course.course_code;
  const generalElective = course.gpa_general_elective
    || (policy.general_elective_codes || []).includes(code)
    || [course.course_category, course.courseCategory, course.category_path, course.categoryPath,
      course.course_subcategory, course.category]
      .some(value => String(value || '').includes('通识选修'));
  if (generalElective) return '通识选修类不计入绩点';
  const scale = policy.grading_scales?.[code] || course.gradingScale || course.grading_scale;
  return String(scale || '').trim() === '两级制' ? '两级制不计入绩点' : '';
};

export const summarizeGpa = (courses, policy = {}) => {
  const included = courses.filter(course => (
    !gpaExclusionReason(course, policy)
    && course.gpa !== null && course.gpa !== undefined && course.gpa !== ''
    && Number.isFinite(Number(course.gpa)) && Number(course.gpa) >= 0
    && Number.isFinite(Number(course.credit)) && Number(course.credit) > 0
  ));
  const credits = included.reduce((sum, course) => sum + Number(course.credit), 0);
  const points = included.reduce((sum, course) => sum + Number(course.gpa) * Number(course.credit), 0);
  return { average: credits > 0 ? points / credits : null, credits, points, count: included.length };
};

export const calculateGpaImpacts = (courses, policy = {}) => {
  const { credits, points, average } = summarizeGpa(courses, policy);
  return courses.map(course => {
    if (!summarizeGpa([course], policy).count) {
      return { ...course, mean_adjust_delta: 0, exclude_delta: 0 };
    }
    const credit = Number(course.credit);
    const gpa = Number(course.gpa);
    return {
      ...course,
      mean_adjust_delta: credits > 0 ? credit * (gpa - average) / credits : 0,
      exclude_delta: credits > credit ? average - (points - credit * gpa) / (credits - credit) : null,
    };
  });
};

File: backend/core/academic/gpa_policy.py
"""Account-owned GPA preferences and cache-only calculation context."""

from __future__ import annotations

import hashlib
import json
import math
import os
from pathlib import Path
import tempfile
import threading

from backend.core.cache import CacheKey
from backend.core.runtime.config import secure_file


LEGACY = "through_2024"
MODERN = "from_2025"
MODES = (LEGACY, MODERN)


def default_mode(account: str) -> str:
    prefix = str(account)[:4]
    return MODERN if len(prefix) == 4 and prefix.isascii() and prefix.isdigit() and int(prefix) >= 2025 else LEGACY


def general_elective(course: dict) -> bool:
    return bool(course.get("gpa_general_elective")) or any(
        "通识选修" in str(course.get(field) or "")
        for field in ("course_category", "category_path", "course_subcategory")
    )


def calculate_gpa(courses: list[dict], context: dict) -> dict:
    excluded = set(context.get("general_elective_codes") or [])
    scales = context.get("grading_scales") or {}
    points = credits = 0.0
    count = 0
    for course in courses:
        code = str(course.get("code") or course.get("course_code") or "")
        if context["mode"] == MODERN and (
            general_elective(course) or code in excluded
            or str(scales.get(code) or course.get("grading_scale") or "").strip() == "两级制"
        ):
            continue
        try:
            credit, gpa = float(course["credit"]), float(course["gpa"])
        except (KeyError, ValueError, TypeError):
            continue
        if not math.isfinite(credit) or not math.isfinite(gpa) or credit <= 0 or gpa < 0:
            continue
        points += credit * gpa
        credits += credit
        count += 1
    return {"average": points / credits if credits else None, "credits": credits, "count": count}


class GpaPolicyService:
    def __init__(self, data_dir, store, registry=None):
        self.root = Path(data_dir) / "gpa_preferences"
        self.store = store
        self.registry = registry
        self._lock = threading.RLock()

    def _payload(self, entry, resource):
        if not entry or not isinstance(entry.payload, dict):
            return {}
        if self.registry:
            spec = self.registry.get(resource)
            if (entry.schema_version != spec.schema_version
                    or entry.revision_algorithm_version != spec.revision_algorithm_version
                    or entry.payload_type != spec.payload_type):
                return {}
        return entry.payload

    def _path(self, account):
        return self.root / (hashlib.sha256(str(account).encode()).hexdigest() + ".json")

    def preference(self, account):
        with self._lock:
            try:
                saved = json.loads(self._path(account).read_text(encoding="utf-8"))
                override = saved.get("mode")
            except (OSError, ValueError, AttributeError):
                override = None
        override = override if override in MODES else None
        return {"mode": override or default_mode(account), "default_mode": default_mode(account), "override": override}

    def save(self, account, mode):
        if mode is not None and mode not in MODES:
            raise ValueError("invalid GPA policy")
        with self._lock:
            self.root.mkdir(parents=True, exist_ok=True)
            fd, name = tempfile.mkstemp(prefix=".gpa-", suffix=".tmp", dir=self.root)
            try:
                with os.fdopen(fd, "w", encoding="utf-8") as handle:
                    json.dump({"mode": mode}, handle)
                    handle.flush()
                    os.fsync(handle.fileno())
                secure_file(Path(name))
                os.replace(name, self._path(account))
            finally:
                if os.path.exists(name):
                    os.unlink(name)
        return self.context(account)

    def context(self, account, scores=None):
        result = self.preference(account)
        report = self.store.get(CacheKey(account, "academic-report"))
        payload = self._payload(report, "academic-report")
        payload = payload.get("report", payload)
        codes, excluded = set(), set()

        def walk(nodes, inherited=False):
            for node in nodes or []:
                blocked = inherited or "通识选修" in str(node.get("name") or "")
                for course in node.get("courses") or []:
                    code = str(course.get("course_code") or course.get("code") or "")
                    if code:
                        codes.add(code)
                        if blocked or general_elective(course):
                            excluded.add(code)
                walk(node.get("children"), blocked)

        walk(payload.get("categories"))
        if scores is None:
            scores_entry = self.store.get(CacheKey(account, "scores"))
            scores = self._payload(scores_entry, "scores").get("scores") or []
        for course in scores:
            code = str(course.get("code") or "")
            if code:
                codes.add(code)
                if general_elective(course):
                    excluded.add(code)
        scales = {}
        entries = self.store.get_many(
            CacheKey(account, "course-outline-metadata", f"course:{code}") for code in codes
        )
        for code in codes:
            entry = entries.get(CacheKey(account, "course-outline-metadata", f"course:{code}"))
            metadata = self._payload(entry, "course-outline-metadata")
            if metadata.get("grading_scale"):
                scales[code] = metadata["grading_scale"]
        unknown = sum(
            1 for course in scores
            if str(course.get("code") or "") not in excluded
            and not (scales.get(str(course.get("code") or "")) or course.get("grading_scale"))
        )
        return {
            **result, "general_elective_codes": sorted(excluded), "grading_scales": scales,
            "report_available": bool(payload.get("categories")), "missing_grading_scales": unknown,
        }

    def summarize(self, account, courses):
        context = self.context(account, courses)
        return {**calculate_gpa(courses, context), "policy": context}
